### ISSUE TITLE
How to generate the pictures in the paper (Figure 4, 5, 6)

### ISSUE THREAD

**Issue by gt980103** [hamidkarimi/DeepDIG#1]
And can DeepDIG use the ImageNet dataset？

**Comment by hamidkarimi** [comment]
In theory yes though I have not tried it. One note here is perhaps you need to have sophisticated autoencoders for imagenet

**Comment by gt980103** [comment]
> In theory yes though I have not tried it. One note here is perhaps you need to have sophisticated autoencoders for imagenet

Thank you for your reply, but how to get Figure 4, 5, and 6 in the paper by running the code
I only got some visualizations of the generated samples

**Comment by hamidkarimi** [comment]
> > In theory yes though I have not tried it. One note here is perhaps you need to have sophisticated autoencoders for imagenet
> 
> Thank you for your reply, but how to get Figure 4, 5, and 6 in the paper by running the code
> I only got some visualizations of the generated samples

I have not included the code for that part. But it is quite easy and  I used PCA to map samples to 2D.

1. pass the images of digits '1' and '2' to the pre-trained model and get their features (embeddings). Do this for both test and train images. Essentially, as written in the code, you need to get four sets of features.
2. Train a PCA with 2 components on features of train sets (see the code below). Infer the 2-D transformed features for test samples (i.e., pca.tranform())
3. pass the image of gendered borderline examples and get their features too. You should note that for this, as explained in the paper, there are two sets of borderline examples '1'-->'2' and '2'-->'1' . Similar to test samples, use the fitted PCA and transform the features of borderline examples to 2-D. concatenate the 2D vectors of both types of borderline examples.
4. Plot 2D samples on a scatter plot.

**The code is not clean but hopefully you can get the idea and use it.** 


    def draw_pca():
        from sklearn.decomposition import PCA
        import matplotlib.pyplot as plt
        plt.rcParams["figure.figsize"] = (16,10)
        #font = {'family' : 'normal',
        #        'size'   : 16}
        #plt.rc('font', **font)
        pre_trained_model =torch.load(args.project_dir+'PreTrainedModels/'+args.dataset+'/'
                                      + args.pre_trained_model+'/pre_trained_model.m').to(device)
    
        pre_trained_model.eval()
    
        train_data_s, _ = dataset_utils.get_class_specific_data(Class=classes['s'], split='train')
        train_data_s,_ = get_pre_trained_model_features_predictions(train_data_s,pre_trained_model,get_probs=False)
        train_data_s = train_data_s.reshape(train_data_s.shape[0],-1)
    
    
        test_data_s, _ = dataset_utils.get_class_specific_data(Class=classes['s'], split='test')
        test_data_s,_ = get_pre_trained_model_features_predictions(test_data_s,pre_trained_model,get_probs=False)
        test_data_s = test_data_s.reshape(test_data_s.shape[0],-1)
    
    
        train_data_t, _ = dataset_utils.get_class_specific_data(Class=classes['t'], split='train')
        train_data_t,_ = get_pre_trained_model_features_predictions(train_data_t,pre_trained_model,get_probs=False)
        train_data_t = train_data_t.reshape(train_data_t.shape[0],-1)
    
        test_data_t, _ = dataset_utils.get_class_specific_data(Class=classes['t'], split='test')
        test_data_t,_ = get_pre_trained_model_features_predictions(test_data_t,pre_trained_model,get_probs=False)
        test_data_t = test_data_t.reshape(test_data_t.shape[0],-1)
    
    
        pca = PCA(n_components=2)
        pca.fit(np.concatenate((train_data_s,train_data_t)))
        train_data_s = pca.transform(train_data_s)
        train_data_t = pca.transform(train_data_t)
        test_data_s = pca.transform(test_data_s)
        test_data_t=pca.transform(test_data_t)
    
        if args.dataset=='MNIST':
            marker1 = 'o'
            marker2 = 'o'
            marker1_test1='+'
            marker1_test2='+'
            label1="'1'"
            label2="'2'"
        elif args.dataset=='FASHIONMNIST':
            marker1 = 'o'
            marker2 = 'o'
            marker1_test1='+'
            marker1_test2='+'
            label1="'Trouser'"
            label2="'Pullover'"
        else:
            marker1 = 'o'
            marker2 = 'o'
            marker1_test1='+'
            marker1_test2='+'
            label1="'Automobile'"
            label2="'Bird'"
    
        plt.scatter(train_data_s[:, 0], train_data_s[:, 1],
                        marker=marker1, color='b',  label="Train {}".format(label1))#cmap=plt.cm.coolwarm
        plt.scatter(test_data_s[:, 0], test_data_s[:, 1],
                        marker=marker1_test1, color='r', label="Test {}".format(label1))
    
        plt.scatter(train_data_t[:, 0], train_data_t[:, 1],
                        marker=marker2, color='g', label="Train {}".format(label2)) #r"${}$".format(marker2)
    
        plt.scatter(test_data_t[:, 0], test_data_t[:, 1],
                        marker=marker1_test2, color='k', label="Test {}".format(label2))
    
        d = pickle.load(open(class_s_t_dir+'borderline_deepdig_{}.pkl'.format(args.middle_point_threshold),'rb'))
        borderline_s_t = d['success_samples']['samples']
        borderline_s_t,_ = get_pre_trained_model_features_predictions(borderline_s_t,pre_trained_model,get_probs=False)
        borderline_s_t=borderline_s_t.reshape(borderline_s_t.shape[0],-1)
        borderline_s_t = pca.transform(borderline_s_t)
    
        d = pickle.load(open(class_t_s_dir+'borderline_deepdig_{}.pkl'.format(args.middle_point_threshold),'rb'))
        borderline_t_s = d['success_samples']['samples']
        borderline_t_s,_ = get_pre_trained_model_features_predictions(borderline_t_s,pre_trained_model,get_probs=False)
        borderline_t_s=borderline_t_s.reshape(borderline_t_s.shape[0],-1)
        borderline_t_s = pca.transform(borderline_t_s)
    
        plt.scatter(np.concatenate((borderline_s_t[:, 0],borderline_t_s[:,0])),     np.concatenate((borderline_s_t[:, 1],borderline_t_s[:,1])),
                                  marker='h', color='y', label="Borderline")
    
        plt.axis('off')
        legend_properties = {'weight':'bold','size':20}
        plt.legend(ncol=1,loc='upper right',prop=legend_properties,markerscale=2)
    
        plt.show()
        return

    def get_pre_trained_model_features_predictions(data,pre_trained_model,get_probs=False):
        features = []
        predictions = []
        probs=[]
        with torch.no_grad():
            for i in range(0,len(data),args.batch_size):
                start = i
                end = min(start+args.batch_size,len(data))
                Z,feats = pre_trained_model(data[start:end].to(device))
                probs.append(F.softmax(Z, dim=1).cpu().numpy())
                predictions.append(np.argmax(F.softmax(Z, dim=1).cpu().numpy(),axis=1))
                feats = feats.cpu().numpy()
                features.append(feats)
            probs = np.concatenate(probs)
            features = np.concatenate(features)
            predictions = np.concatenate(predictions)
        if get_probs:
            return features, probs
        return features,predictions

**Comment by gt980103** [comment]
> > > In theory yes though I have not tried it. One note here is perhaps you need to have sophisticated autoencoders for imagenet
> > 
> > 
> > Thank you for your reply, but how to get Figure 4, 5, and 6 in the paper by running the code
> > I only got some visualizations of the generated samples
> 
> I have not included the code for that part. But it is quite easy and I used PCA to map samples to 2D.
> 
> 1. pass the images of digits '1' and '2' to the pre-trained model and get their features (embeddings). Do this for both test and train images. Essentially, as written in the code, you need to get four sets of features.
> 2. Train a PCA with 2 components on features of train sets (see the code below). Infer the 2-D transformed features for test samples (i.e., pca.tranform())
> 3. pass the image of gendered borderline examples and get their features too. You should note that for this, as explained in the paper, there are two sets of borderline examples '1'-->'2' and '2'-->'1' . Similar to test samples, use the fitted PCA and transform the features of borderline examples to 2-D. concatenate the 2D vectors of both types of borderline examples.
> 4. Plot 2D samples on a scatter plot.
> 
> **The code is not clean but hopefully you can get the idea and use it.**
> 
> ```
> def draw_pca():
>     from sklearn.decomposition import PCA
>     import matplotlib.pyplot as plt
>     plt.rcParams["figure.figsize"] = (16,10)
>     #font = {'family' : 'normal',
>     #        'size'   : 16}
>     #plt.rc('font', **font)
>     pre_trained_model =torch.load(args.project_dir+'PreTrainedModels/'+args.dataset+'/'
>                                   + args.pre_trained_model+'/pre_trained_model.m').to(device)
> 
>     pre_trained_model.eval()
> 
>     train_data_s, _ = dataset_utils.get_class_specific_data(Class=classes['s'], split='train')
>     train_data_s,_ = get_pre_trained_model_features_predictions(train_data_s,pre_trained_model,get_probs=False)
>     train_data_s = train_data_s.reshape(train_data_s.shape[0],-1)
> 
> 
>     test_data_s, _ = dataset_utils.get_class_specific_data(Class=classes['s'], split='test')
>     test_data_s,_ = get_pre_trained_model_features_predictions(test_data_s,pre_trained_model,get_probs=False)
>     test_data_s = test_data_s.reshape(test_data_s.shape[0],-1)
> 
> 
>     train_data_t, _ = dataset_utils.get_class_specific_data(Class=classes['t'], split='train')
>     train_data_t,_ = get_pre_trained_model_features_predictions(train_data_t,pre_trained_model,get_probs=False)
>     train_data_t = train_data_t.reshape(train_data_t.shape[0],-1)
> 
>     test_data_t, _ = dataset_utils.get_class_specific_data(Class=classes['t'], split='test')
>     test_data_t,_ = get_pre_trained_model_features_predictions(test_data_t,pre_trained_model,get_probs=False)
>     test_data_t = test_data_t.reshape(test_data_t.shape[0],-1)
> 
> 
>     pca = PCA(n_components=2)
>     pca.fit(np.concatenate((train_data_s,train_data_t)))
>     train_data_s = pca.transform(train_data_s)
>     train_data_t = pca.transform(train_data_t)
>     test_data_s = pca.transform(test_data_s)
>     test_data_t=pca.transform(test_data_t)
> 
>     if args.dataset=='MNIST':
>         marker1 = 'o'
>         marker2 = 'o'
>         marker1_test1='+'
>         marker1_test2='+'
>         label1="'1'"
>         label2="'2'"
>     elif args.dataset=='FASHIONMNIST':
>         marker1 = 'o'
>         marker2 = 'o'
>         marker1_test1='+'
>         marker1_test2='+'
>         label1="'Trouser'"
>         label2="'Pullover'"
>     else:
>         marker1 = 'o'
>         marker2 = 'o'
>         marker1_test1='+'
>         marker1_test2='+'
>         label1="'Automobile'"
>         label2="'Bird'"
> 
>     plt.scatter(train_data_s[:, 0], train_data_s[:, 1],
>                     marker=marker1, color='b',  label="Train {}".format(label1))#cmap=plt.cm.coolwarm
>     plt.scatter(test_data_s[:, 0], test_data_s[:, 1],
>                     marker=marker1_test1, color='r', label="Test {}".format(label1))
> 
>     plt.scatter(train_data_t[:, 0], train_data_t[:, 1],
>                     marker=marker2, color='g', label="Train {}".format(label2)) #r"${}$".format(marker2)
> 
>     plt.scatter(test_data_t[:, 0], test_data_t[:, 1],
>                     marker=marker1_test2, color='k', label="Test {}".format(label2))
> 
>     d = pickle.load(open(class_s_t_dir+'borderline_deepdig_{}.pkl'.format(args.middle_point_threshold),'rb'))
>     borderline_s_t = d['success_samples']['samples']
>     borderline_s_t,_ = get_pre_trained_model_features_predictions(borderline_s_t,pre_trained_model,get_probs=False)
>     borderline_s_t=borderline_s_t.reshape(borderline_s_t.shape[0],-1)
>     borderline_s_t = pca.transform(borderline_s_t)
> 
>     d = pickle.load(open(class_t_s_dir+'borderline_deepdig_{}.pkl'.format(args.middle_point_threshold),'rb'))
>     borderline_t_s = d['success_samples']['samples']
>     borderline_t_s,_ = get_pre_trained_model_features_predictions(borderline_t_s,pre_trained_model,get_probs=False)
>     borderline_t_s=borderline_t_s.reshape(borderline_t_s.shape[0],-1)
>     borderline_t_s = pca.transform(borderline_t_s)
> 
>     plt.scatter(np.concatenate((borderline_s_t[:, 0],borderline_t_s[:,0])),     np.concatenate((borderline_s_t[:, 1],borderline_t_s[:,1])),
>                               marker='h', color='y', label="Borderline")
> 
>     plt.axis('off')
>     legend_properties = {'weight':'bold','size':20}
>     plt.legend(ncol=1,loc='upper right',prop=legend_properties,markerscale=2)
> 
>     plt.show()
>     return
> 
> def get_pre_trained_model_features_predictions(data,pre_trained_model,get_probs=False):
>     features = []
>     predictions = []
>     probs=[]
>     with torch.no_grad():
>         for i in range(0,len(data),args.batch_size):
>             start = i
>             end = min(start+args.batch_size,len(data))
>             Z,feats = pre_trained_model(data[start:end].to(device))
>             probs.append(F.softmax(Z, dim=1).cpu().numpy())
>             predictions.append(np.argmax(F.softmax(Z, dim=1).cpu().numpy(),axis=1))
>             feats = feats.cpu().numpy()
>             features.append(feats)
>         probs = np.concatenate(probs)
>         features = np.concatenate(features)
>         predictions = np.concatenate(predictions)
>     if get_probs:
>         return features, probs
>     return features,predictions
> ```

With this, I'll try to understand the whole project. Thanks for your guidance, it really helps me a lot.
